### DISCUSSION TITLE
6장 프론트엔드 배포와 관련된 질문입니다. (p.381)

### DISCUSSION THREAD

**Discussion by jungboke** [fsoftwareengineer/todo-application#96] [질문과 답변] [answered]
p.381의 실습코드 6-17 eb create를 하면서 빈스톡 환경을 구축하는데, 502 page 에러가 발생합니다.
로그를 확인하고, 이와 관련된 다른 사람들의 질문과 해결책을 봐도 해결이 되지 않습니다.

package.json이나 config.yml은 github의 소스코드를 그대로 사용했습니다. Node의 버전은 최신이나, 이는 영향을
끼치지 않은듯합니다.

**Comment by jungboke** [comment]
2022/03/07 06:22:03 [error] 588#588: *426 connect() failed (111: Connection refused) while connecting to upstream, client: 172.31.21.38, server: , request: "GET / HTTP/1.1", upstream: "http://127.0.0.1:8081/", host: "172.31.3.164"
2022/03/07 06:22:04 [error] 588#588: *428 connect() failed (111: Connection refused) while connecting to upstream, client: 172.31.10.205, server: , request: "GET / HTTP/1.1", upstream: "http://127.0.0.1:8081/", host: "172.31.3.164"
2022/03/07 06:22:18 [error] 588#588: *430 connect() failed (111: Connection refused) while connecting to upstream, client: 172.31.21.38, server: , request: "GET / HTTP/1.1", upstream: "http://127.0.0.1:8081/", host: "172.31.3.164"
2022/03/07 06:22:20 [error] 588#588: *432 connect() failed (111: Connection refused) while connecting to upstream, client: 172.31.10.205, server: , request: "GET / HTTP/1.1", upstream: "http://127.0.0.1:8081/", host: "172.31.3.164"
2022/03/07 06:22:33 [error] 588#588: *434 connect() failed (111: Connection refused) while connecting to upstream, client: 172.31.21.38, server: , request: "GET / HTTP/1.1", upstream: "http://127.0.0.1:8081/", host: "172.31.3.164"
2022/03/07 06:22:35 [error] 588#588: *436 connect() failed (111: Connection refused) while connecting to upstream, client: 172.31.10.205, server: , request: "GET / HTTP/1.1", upstream: "http://127.0.0.1:8081/", host: "172.31.3.164"
2022/03/07 06:22:48 [error] 588#588: *438 connect() failed (111: Connection refused) while connecting to upstream, client: 172.31.21.38, server: , request: "GET / HTTP/1.1", upstream: "http://127.0.0.1:8081/", host: "172.31.3.164"
2022/03/07 06:22:50 [error] 588#588: *440 connect() failed (111: Connection refused) while connecting to upstream, client: 172.31.10.205, server: , request: "GET / HTTP/1.1", upstream: "http://127.0.0.1:8081/", host: "172.31.3.164"
2022/03/07 06:23:03 [error] 588#588: *442 connect() failed (111: Connection refused) while connecting to upstream, client: 172.31.21.38, server: , request: "GET / HTTP/1.1", upstream: "http://127.0.0.1:8081/", host: "172.31.3.164"
2022/03/07 06:23:05 [error] 588#588: *444 connect() failed (111: Connection refused) while connecting to upstream, client: 172.31.10.205, server: , request: "GET / HTTP/1.1", upstream: "http://127.0.0.1:8081/", host: "172.31.3.164"
2022/03/07 06:23:18 [error] 588#588: *446 connect() failed (111: Connection refused) while connecting to upstream, client: 172.31.21.38, server: , request: "GET / HTTP/1.1", upstream: "http://127.0.0.1:8081/", host: "172.31.3.164"
2022/03/07 06:23:20 [error] 588#588: *448 connect() failed (111: Connection refused) while connecting to upstream, client: 172.31.10.205, server: , request: "GET / HTTP/1.1", upstream: "http://127.0.0.1:8081/", host: "172.31.3.164"
2022/03/07 06:23:33 [error] 588#588: *450 connect() failed (111: Connection refused) while connecting to upstream, client: 172.31.21.38, server: , request: "GET / HTTP/1.1", upstream: "http://127.0.0.1:8081/", host: "172.31.3.164"
2022/03/07 06:23:35 [error] 588#588: *452 connect() failed (111: Connection refused) while connecting to upstream, client: 172.31.10.205, server: , request: "GET / HTTP/1.1", upstream: "http://127.0.0.1:8081/", host: "172.31.3.164"

/var/log/nginx/error.log에서 확인해보니 Connection refused 에러도 보입니다. 해당 EB의 환경변수로 PORT:8081을 주었고, 애플리케이션의 PORT도 8081로 맞춰줬습니다

[web.stdout.log](https://github.com/fsoftwareengineer/todo-application/files/8195656/web.stdout.log)

해당 로그 파일을 보면 
Mar  7 06:52:31 ip-172-31-10-87 web: > todo-react-app@0.1.0 start /var/app/current
Mar  7 06:52:31 ip-172-31-10-87 web: > set PORT=8080 && react-scripts start
Mar  7 06:52:33 ip-172-31-10-87 web: Could not find a required file.
Mar  7 06:52:33 ip-172-31-10-87 web: Name: index.html
Mar  7 06:52:33 ip-172-31-10-87 web: Searched in: /var/app/current/public
Mar  7 06:52:33 ip-172-31-10-87 web: npm ERR! code ELIFECYCLE
Mar  7 06:52:33 ip-172-31-10-87 web: npm ERR! errno 1
Mar  7 06:52:33 ip-172-31-10-87 web: npm ERR! todo-react-app@0.1.0 start: `set PORT=8080 && react-scripts start`
Mar  7 06:52:33 ip-172-31-10-87 web: npm ERR! Exit status 1
Mar  7 06:52:33 ip-172-31-10-87 web: npm ERR!
Mar  7 06:52:33 ip-172-31-10-87 web: npm ERR! Failed at the todo-react-app@0.1.0 start script.
Mar  7 06:52:33 ip-172-31-10-87 web: npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
Mar  7 06:52:33 ip-172-31-10-87 web: npm ERR! A complete log of this run can be found in:
Mar  7 06:52:33 ip-172-31-10-87 web: npm ERR!     /home/webapp/.npm/_logs/2022-03-07T06_52_33_936Z-debug.log

Could not find a required file이 뜨고 있습니다. 분명 로컬환경에서는 문제없이 start되는데 일라스틱 빈스톡을 통해 인스턴스에
로드되면 해당 에러가 발생합니다. 
.

**Comment by fsoftwareengineer** [comment]
public directory에 index.html이 있는지 확인 해 주시기 바랍니다. 또 git을 사용하신다면 index.html이 commit되어 있는지 확인 해 주시기 바랍니다.ㅁ

**Comment by jungboke** [comment, marked as answer]
위의 Could not find a required file 에러는 package.json 의 dependency중 react-script의 버전 문제라는 것을 알았습니다.
따라서 해결하고 다시 빈스톡을 돌리니까 React 초기화면만 뜨는 오류가 생겨 .git을 삭제하고 다시 빈스톡을 재구축해 해결했습니다.
마지막 문제로 sh: /var/app/current/node_modules/.bin/react-scripts: Permission denied가 발생해서 다시 502 gateway가 뜹니다.

https://github.com/fsoftwareengineer/todo-application/discussions/35#discussioncomment-1888637

위의 방법대로 setfacl -R -m u:ec2-user:rwx node_modules 을 직접 인스턴스에 들어가 적용시켰는데, 이후 sudo service nginx restart를 하면 /var/log/web.stdout.log에 로그가 추가되지 않는걸 보아, nginx가 돌아가지 않는 것 같습니다.
.npmrc로 unsafe-perm=true를 추가해도 효과가 없었습니다. 

sudo service nginx restart가 동작하지 않아 npm run start&을 실행해주니 인스턴스가 앱이 정상적으로 동작했습니다!

마지막으로 궁금한게 있는데, 빈스톡이 인스턴스 내에서 리액트 앱을 실행시킬 때, npm run start를 사용해서 실행 시키는건가요?
또한, 해당 npm run start는 build 폴더와 연관돼있는건가요?

**Reply by fsoftwareengineer** [reply]
처음 일라스틱빈스톡에서 Managed Platform을 Node.js로 선택하셨나요?
